cirq-core/cirq/qis/clifford_tableau.py

```diff
@@ -176,6 +178,43 @@ def _str_full_(self) -> str:
 
         return string
 
+    def merged_with(self, second: 'CliffordTableau') -> 'CliffordTableau':
+        """Returns a merged CliffordTableau of self tableau and second tableau.
+
+        The meaning of merge two clifford tableau is the corresponding unitary operation
+        of merged clifford tableau is equal to (up to global phase) the composed unitary
+        operation of self tableau and the one of second tableau.
+        """
+        m1 = self.matrix().astype(int)
+        m2 = second.matrix().astype(int)
+
+        p1 = self.rs.astype(int)
+        p2 = second.rs.astype(int)
+
+        merged_m = np.mod(m1.dot(m2), 2)
+        phase = np.mod(p1 + m1.dot(p2), 2)
+
+        # we need more phase correction for expanding Y to XZ and swapping Z_iX_i order.
+        for k in range(2 * self.n):
+            swap_phase = 0  # value betwen 0 and 3 representing [1, i, -1, -i] respectively.
+            prev_row_sum = np.zeros([2 * self.n])
+            swap_phase += np.sum(m1[k, : self.n] * m1[k, self.n :])  # Y gate => iXZ
+            for i, v in enumerate(m1[k]):
```

Comment:
Can this be done with a vectorized numpy operation? My "slow python iteration over numpy arrays" senses are tingling.

If you need to only do it where the entries are not equal to zero, you can do fancy indexing:

```python
non_zero = m1[k] != 0
swap_phase += np.sum(m2[non_zero, :, self.n] * m2, [non_zero] ....
```

Reply:
Unfortunately, that is not doable because of this line `prev_row_sum += m2[i]`. It is a stateful for-loop. If we indeed worry about the speed, we can consider Cython (NumPy is dependent on it anyway).

Reply:
thanks